src/base/net/downloadhandler.cpp

```diff
@@ -172,38 +170,35 @@ void Net::DownloadHandler::handleRedirection(const QUrl &newUrl)
     if (newUrlString.startsWith("magnet:", Qt::CaseInsensitive)) {
         qDebug("Magnet redirect detected.");
         m_reply->abort();
-        if (m_downloadRequest.handleRedirectToMagnet())
-            emit redirectedToMagnet(m_downloadRequest.url(), newUrlString);
-        else
-            emit downloadFailed(m_downloadRequest.url(), tr("Unexpected redirect to magnet URI."));
+        m_result.status = DownloadStatus::RedirectedToMagnet;
+        m_result.magnet = newUrlString;
 
-        this->deleteLater();
+        finish();
         return;
     }
 
     DownloadHandler *redirected = m_manager->download(DownloadRequest(m_downloadRequest).url(newUrlString));
     redirected->m_redirectionCounter = (m_redirectionCounter + 1);
-    connect(redirected, &DownloadHandler::destroyed, this, &DownloadHandler::deleteLater);
-    connect(redirected, &DownloadHandler::downloadFailed, this, [this](const QString &, const QString &reason)
-    {
-        emit downloadFailed(url(), reason);
-    });
-    connect(redirected, &DownloadHandler::redirectedToMagnet, this, [this](const QString &, const QString &magnetUri)
-    {
-        emit redirectedToMagnet(url(), magnetUri);
-    });
-    connect(redirected, static_cast<void (DownloadHandler::*)(const QString &, const QString &)>(&DownloadHandler::downloadFinished)
-            , this, [this](const QString &, const QString &fileName)
+    connect(redirected, &DownloadHandler::finished, this, [this](const DownloadResult &result)
     {
-        emit downloadFinished(url(), fileName);
-    });
-    connect(redirected, static_cast<void (DownloadHandler::*)(const QString &, const QByteArray &)>(&DownloadHandler::downloadFinished)
-            , this, [this](const QString &, const QByteArray &data)
-    {
-        emit downloadFinished(url(), data);
+        m_result = result;
+        m_result.url = url();
```

Comment:
I think `struct DownloadResult` should have 2 url fields:
1. original/start url, this is mainly for error reporting, otherwise after a number of redirections it will be very unclear what was the start url.
2. final url, this is the destination.

I hope I'm not mistaken that the current pr will lose the original/start url information.

Reply:
It provides only start URL. Redirected URL is never used by client so we don't have it in result.